File: 1-10/4. Median of Two Sorted Arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function (nums1 = [1, 3], nums2 = [2]) {
    let ans = [];
    let i = 0,
        j = 0;
    while (i < nums1.length || j < nums2.length) {
        if (i == nums1.length) {
            ans = ans.concat(nums2.slice(j));
            break;
        }
        if (j == nums2.length) {
            ans = ans.concat(nums1.slice(i));
            break;
        }
        if (nums1[i] < nums2[j]) {
            ans.push(nums1[i]);
            i++;
        } else {
            ans.push(nums2[j]);
            j++;
        }
    }
    let len = ans.length;
    if (len % 2 == 0) {
        return (ans[len / 2 - 1] + ans[len / 2]) / 2;
    } else {
        return ans[len / 2 - 0.5];
    }
};

File: 1-10/4. Median of Two Sorted Arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A, B = nums1, nums2
        total = len(nums1) + len(nums2)
        half = total // 2
        # use smaller array as the based array
        # otherwise j will be out of bounds
        # i will always be non-negative, but j will be negative if based array is too large
        # so that j + 1 < 0
        if len(A) > len(B):
            A, B = B, A

        l, r = 0, len(A) - 1

        while True:
            i = l + (r - l) // 2
            j = half - i - 2

            Aleft = A[i] if i >= 0 else float("-inf")
            Aright = A[i + 1] if (i + 1) < len(A) else float("inf")
            Bleft = B[j] if j >= 0 else float("-inf")
            Bright = B[j + 1] if (j + 1) < len(B) else float("inf")

            if Aleft <= Bright and Bleft <= Aright:
                if total % 2:
                    return min(Aright, Bright)
                else:
                    return (max(Aleft, Bleft) + min(Aright, Bright)) / 2

            elif Aleft > Bright:
                r = i - 1
            else:
                l = i + 1
